examples/example-with-encryption/main.tf

```diff
@@ -0,0 +1,140 @@
+# conditionals in this template to fail.
+terraform {
+  required_version = ">= 0.9.3, != 0.9.5"
+}
+
+# ---------------------------------------------------------------------------------------------------------------------
+# DEPLOY THE CONSUL SERVER NODES
+# ---------------------------------------------------------------------------------------------------------------------
+
+module "consul_servers" {
+  # When using these modules in your own templates, you will need to use a Git URL with a ref attribute that pins you
+  # to a specific version of the modules, such as the following example:
+  # source = "git::git@github.com:hashicorp/terraform-aws-consul.git//modules/consul-cluster?ref=v0.0.1"
+  source = "../../modules/consul-cluster"
+
+  cluster_name  = "${var.cluster_name}-server"
+  cluster_size  = "${var.num_servers}"
+  instance_type = "t2.micro"
+  spot_price    = "${var.spot_price}"
+
+  # The EC2 Instances will use these tags to automatically discover each other and form a cluster
+  cluster_tag_key   = "${var.cluster_tag_key}"
+  cluster_tag_value = "${var.cluster_name}"
+
+  ami_id    = "${var.ami_id}"
+  user_data = "${data.template_file.user_data_server.rendered}"
+
+  vpc_id     = "${data.aws_vpc.default.id}"
+  subnet_ids = "${data.aws_subnet_ids.default.ids}"
+
+  # To make testing easier, we allow Consul and SSH requests from any IP address here but in a production
+  # deployment, we strongly recommend you limit this to the IP address ranges of known, trusted servers inside your VPC.
+  allowed_ssh_cidr_blocks = ["0.0.0.0/0"]
+
+  allowed_inbound_cidr_blocks = ["0.0.0.0/0"]
+  ssh_key_name                = "${var.ssh_key_name}"
+
+  tags = [
+    {
+      key                 = "Environment"
+      value               = "development"
+      propagate_at_launch = true
+    },
+  ]
+}
+
+# ---------------------------------------------------------------------------------------------------------------------
+# THE USER DATA SCRIPT THAT WILL RUN ON EACH CONSUL SERVER EC2 INSTANCE WHEN IT'S BOOTING
+# This script will configure and start Consul
+# ---------------------------------------------------------------------------------------------------------------------
+
+data "template_file" "user_data_server" {
+  template = "${file("${path.module}/user-data-server.sh")}"
+
+  vars {
+    cluster_tag_key   = "${var.cluster_tag_key}"
+    cluster_tag_value = "${var.cluster_name}"
+    gossip_encryption_configuration = "${var.enable_gossip_encryption && var.gossip_encryption_key != "" ? "--enable-gossip-encryption --gossip-encryption-key ${var.gossip_encryption_key}" : ""}"
+    rpc_encryption_configuration = "${var.enable_rpc_encryption && var.ca_path != "" && var.cert_file_path != "" && var.key_file_path != "" ? "--enable-rpc-encryption --ca-path ${var.ca_path} --cert-file-path ${var.cert_file_path} --key-file-path ${var.key_file_path}" : ""}"
```

Comment:
I'm not a big fan of passing command-line args from Terraform to user data. I think it's a bit cleaner if you *just* pass the parameters straight through:

```hcl
enable_gossip_encryption = "${var.enable_gossip_encryption}"
gossip_encryption_key    = "${var.gossip_encryption_key}"

enable_rpc_encryption = "${var.enable_rpc_encryption}"
ca_path               = "${var.ca_path}"

# ... etc ...
```

The User Data script itself can check the values of the parameters and set command line args as necessary.